# W tym dokumencie:
* Elementy pracy z systemem Linux
    * Prawa dostępu
    * Łączenie poleceń w większą całość - pipes
    * Przekierowanie
    * Wyszukiwanie w tekście
    * Archiwizacja danych
    * Procesy i zadania
    * Programowanie w BASH, skrypty

Źródła i literatura:
* https://www.doc.ic.ac.uk/~wjk/UnixIntro/

***

## Prawa dostępu
Ze względu na przeznaczenie do wykorzystania przez wielu użytkowników systemy z rodziny UNIX w zasadzie od początku wprowadzały system kontroli dostępu do plików (coś co w systemach w zamyśle 'osobistych' zaistniało w zasadzie niedawno).

Prawa dostępu kontrolują jakie operacje są dopuszczalne dla użytkowników. Dla każdego pliku zdefiniowane jest jakie operacje mogą zostać wykonane przez trzy kategorie użytkowników. Prawa to: czytanie (r), zapis (w) i wykonanie (x), a kategorie użytkowników to: użytkownik (u), grupa (g) i inni (o). Prawa dostępu może zmienić właściciel (użytkownik) albo super-użytkownik tzw. root (bo może wszystko).

Trochę o znaczeniu praw:
* *r* oznacza prawo do odczytu. W przypadku plików jest to możliwość czytania pliku, w przypadku katalogu wylistowania jego zawartości.
* *w* oznacza prawo do zapisu. W przypadku pliku oznacza prawo do modyfikacji, kasowania. W przypadku katalogu prawo do tworzenia nowych plików w katalogu.
* *x* oznacza prawo do wykonania pliku, zazwyczaj jest to program bądź skrypt powłoki. W przypadku katalogu oznacza prawo wejścia do katalogu, ale nie wylistowania zawartość.

Prawa dostępu można sprawdzić wywołując ```ls``` z opcją ```-l```  
wynik może wyglądać np. tak:  
```ls -l```  
```-rw-rw-r-- 1 sgepner sgepner  110559 mar  5 12:28  burza```  
i oznacza, że użytkownik i grupa mają prawo czytania i modyfikacji, pozostali użytkownicy jedynie czytania.

Do zmiany praw dostępu służy komenda ```chmod```, wywołuje się ją z opcjami, które determinują zmiany praw (*rwx*) dla konkretnych kategorii użytkowników: u,g i o. Np. następujące polecenie
```chmod u+r plik``` nada prawa odczytu pliku jego właścicielowi, natomiast ```chmod g+w,o-r plik``` pozwoli na zapis grupie i uniemożliwi czytanie innym.

Przykład:
Do pliku fig1.png prawo do czytania i pisania mają wszyscy:

In [1]:
ls -l

total 3308
-rw-rw-r-- 1 sgepner sgepner  110559 mar  5 12:28  burza
-rw-r--r-- 1 sgepner sgepner     303 mar  6 12:20  burza_short
-rw-r--r-- 1 sgepner sgepner   13182 mar  3 16:24  fig1.png
-rw-rw-r-- 1 sgepner sgepner   17000 mar  3 16:23  fig1.svg
-rw-rw-r-- 1 sgepner sgepner   16358 mar  3 16:24  fig2.png
-rw-r--r-- 1 sgepner sgepner  419599 mar  6 13:28 'Info 3 - Wykład 1.ipynb'
-rw-r--r-- 1 sgepner sgepner 1291382 mar  6 13:46 'Info 3 - Wykład 1.pdf'
-rw-rw-r-- 1 sgepner sgepner   45703 mar  6 13:46 'Info 3 - Wykład 1.tex'
-rw-r--r-- 1 sgepner sgepner  208330 mar 18 14:31 'Info 3 - Wykład 2.ipynb'
-rw-r--r-- 1 sgepner sgepner   65458 mar  3 16:06  quote-c-is-a-horrible-language-it-s-made-more-horrible-by-the-fact-that-a-lot-of-substandard-linus-torvalds-111-83-18.jpg
-rw-r--r-- 1 sgepner sgepner 1166196 mar  3 17:37  unix.png
-rw-r--r-- 1 sgepner sgepner    6150 mar  3 18:03  Zaznaczenie_004.png
-rw-r--r-- 1 sgepner sgepner    1768 mar  3 18:03  Zaznaczenie_005.png


``` ls -l  ```   
``` -|rwx|rwx|rwx ``` tu wszyscy mają prawo do wszystkiego

Do zmiany praw dostępu służy komenta ```chmod```

In [11]:
ls -l fig1.png

-rw-rw-rw- 1 sgepner sgepner 13182 mar  3 16:24 fig1.png


Użytkownik, chce zabrać prawa innym i ograniczyć prawa grupy:

In [12]:
chmod g-w,o-rw fig1.png

W efekcie:

In [13]:
ls -l fig1.png

-rw-r----- 1 sgepner sgepner 13182 mar  3 16:24 fig1.png


**Ciekawostka**: Prawa dostępu można zapisywać w systemie *ósemkowym*. Wówczas ```rwx``` przedstawić można tak:

* ---  0
* --x  1
* -w-  2
* -wx  3
* r--  4
* r-x  5
* rw-  6
* rwx  7

W takim zapisie zmiana praw pliku *fig1.png* na początkowe wygląda tak:

In [15]:
chmod 666 fig1.png
ls -l fig1.png

-rw-rw-rw- 1 sgepner sgepner 13182 mar  3 16:24 fig1.png


A wcześniej wprowadzone ograniczenie to:

In [16]:
chmod 650 fig1.png
ls -l fig1.png

-rw-r-x--- 1 sgepner sgepner 13182 mar  3 16:24 fig1.png


```chmod``` może działać też w trybie rekurencyjnym, jeżeli wywołany zostanie z flagą ```-r``` zmiany zostaną zastosowane do katalogu i całej zawartości w dół drzewa katalogowego.

### Zmiana właściciela
Do zmiany właściciela pliku służy: ```chown```, natomiast grupę zmienia się przez ```chgrp```:  
```chown [options] [nowy właściciel] file```

Do zmiany właściciela pliku służy: ```chown``` zmiana grupy przez ```chgrp```.

In [17]:
man chown

CHOWN(1)                         User Commands                        CHOWN(1)

NAME
       chown - change file owner and group

SYNOPSIS
       chown [OPTION]... [OWNER][:[GROUP]] FILE...
       chown [OPTION]... --reference=RFILE FILE...

DESCRIPTION
       This manual page documents the GNU version of chown.  chown changes the
       user and/or group ownership of each given file.  If only  an  owner  (a
       user  name or numeric user ID) is given, that user is made the owner of
       each given file, and the files' group is not changed.  If the owner  is
       followed  by  a  colon  and a group name (or numeric group ID), with no
       spaces between them, the group ownership of the  files  is  changed  as
       well.  If a colon but no group name follows the user name, that user is
       made the owner of the files and the group of the files  is  changed  to
       that  user's  login  group.   If the colon and group are given, but the
       owner is omitted, only the gr

## Łączenie poleceń w większą całość (Pipes)
Polecenia powłoki można łączyć w celu uzyskania pożądanego efektu. Tzn. Wynik jednego polecenia stanowić może argument kolejnego. Służy temu operator '''|''' (pipe). Np:

In [134]:
head burza -n 666 | tail -n 3

Witaj, wielki mistrzu!

Witaj! Przybiegam wolę twoją spełnić,


utworzy proces ```head``` i ```tail```. ```head``` przetworzy pierwsze 666 linii pliku burza i przekieruje je jako argument dla ```tail```, które wyświetli 3 ostatnich linii.

Przekierowanie '''|''' będziemy często wykorzystywać.

## Przekierowanie
Poza operatorem ```|``` możemy też przekierowywać wynik działania procesu do pliku. Standardowo polecenia drukują na tzw. standardowe wyjście (ekran). Możemy jednak przekierować to wyjście do dowolnego pliku. W tym celu użyjemy operatorów ``` > ```, ```>>``` i ich wariantów.
* ```>``` - tworzy nowy plik, lub nadpisuje już istniejący
* ```>>```- dopisuje do istniejącego pliku lub tworzy nowy
* ```1>```- przekierowanie wyjścia, bez błędów
* ```2>```- przekierowanie błędów, bez wyjścia
* ```>&```- przekierowanie wyjścia oraz błędów

Przykład: Do pliku zapiszemy date, położenie, listę plików i wyświetlimy zawartość pliku:

In [162]:
date > result_file
pwd >> result_file
ls >> result_file
cat result_file

pią, 20 mar 2020, 17:27:26 CET
/home/sgepner/OneDrive/Dydaktyka/Info_3/Wykład
burza
burza_short
fig1.png
fig1.svg
fig2.png
Info 3 - Wykład 1.aux
Info 3 - Wykład 1.ipynb
Info 3 - Wykład 1.log
Info 3 - Wykład 1.out
Info 3 - Wykład 1.pdf
Info 3 - Wykład 1.synctex.gz
Info 3 - Wykład 1.tex
Info 3 - Wykład 2.ipynb
quote-c-is-a-horrible-language-it-s-made-more-horrible-by-the-fact-that-a-lot-of-substandard-linus-torvalds-111-83-18.jpg
result_file
unix.png
Zaznaczenie_004.png
Zaznaczenie_005.png


Jeżeli powtórzymy pierwszą instrukcję, plik zostanie nadpisany:

In [163]:
date > result_file
cat result_file

pią, 20 mar 2020, 17:28:28 CET


Rozdzielmy jeszcze wynik od błędów:

In [164]:
date 1> result_file 2> /dev/null

**Ciekawostka:** W systemach Unixowych jest specjalny plik ```/dev/null```, reprezentuje on nicość, a wszystko co zostaje na niego przekierowane nie jest zapisywane i przestaje istnieć. Odniesienia do ```/dev/null``` stanowią obecnie kanonem kulturowy: [References in computer culture](https://en.wikipedia.org/wiki/Null_device#References_in_computer_culture). 

## Zabawy z plikami i tekstem
Widzieliśmy już jak wyświetlić zawartość pliku na różne sposoby:
    ``` cat ```, ``` more ```, ``` head ``` czy ``` tail ```.
    
Uzupełnimy teraz arsenał poleceń o ``` file ```, najlepiej niech wypowie się manual:

In [20]:
man file

FILE(1)                   BSD General Commands Manual                  FILE(1)

NAME
     file — determine file type

SYNOPSIS
     file [-bcdEhiklLNnprsvzZ0] [--apple] [--extension] [--mime-encoding]
          [--mime-type] [-e testname] [-F separator] [-f namefile]
          [-m magicfiles] [-P name=value] file ...
     file -C [-m magicfiles]
     file [--help]

DESCRIPTION
     This manual page documents version 5.32 of the file command.

     file tests each argument in an attempt to classify it.  There are three
     sets of tests, performed in this order: filesystem tests, magic tests,
     and language tests.  The first test that succeeds causes the file type to
     be printed.

     The type printed will usually contain one of the words text (the file
     contains only printing characters and a few common control characters and
     is probably safe to read on an ASCII terminal), executable (the file con‐
     tains the result of compiling a program in a form understandable 


     -i, --mime
             Causes the file command to output mime type strings rather than
             the more traditional human readable ones.  Thus it may say
             ‘text/plain; charset=us-ascii’ rather than “ASCII text”.

     --mime-type, --mime-encoding
             Like -i, but print only the specified element(s).

     -k, --keep-going
             Don't stop at the first match, keep going.  Subsequent matches
             will be have the string ‘\012- ’ prepended.  (If you want a new‐
             line, see the -r option.)  The magic pattern with the highest
             strength (see the -l option) comes first.

     -l, --list
             Shows a list of patterns and their strength sorted descending by
             magic(4) strength which is used for the matching (see also the -k
             option).

     -L, --dereference
             option causes symlinks to be followed, as the like-named option
             in ls(1) (on systems that support symbolic links)

           /dev/hda:    application/x-not-regular-file
           /dev/wd0a:   application/x-not-regular-file

HISTORY
     There has been a file command in every UNIX since at least Research
     Version 4 (man page dated November, 1973).  The System V version intro‐
     duced one significant major change: the external list of magic types.
     This slowed the program down slightly but made it a lot more flexible.

     This program, based on the System V version, was written by Ian Darwin
     ⟨ian@darwinsys.com⟩ without looking at anybody else's source code.

     John Gilmore revised the code extensively, making it better than the
     first version.  Geoff Collyer found several inadequacies and provided
     some magic file entries.  Contributions of the ‘&’ operator by Rob McMa‐
     hon, ⟨cudcv@warwick.ac.uk⟩, 1989.

     Guy Harris, ⟨guy@netapp.com⟩, made many changes from 1993 to the present.

     Primary development and maintenance from 1990 to the present by Christos
     

czyli, krótko ```file``` wyświetla informację o pliku, np:

In [118]:
file fig1.png
file fig1.svg
file burza

fig1.png: PNG image data, 388 x 141, 8-bit/color RGBA, non-interlaced
fig1.svg: SVG Scalable Vector Graphics image
burza: UTF-8 Unicode text, with very long lines


Kolejnym przydatnym poleceniem jest ```find```, w najprostszej wersji pozwala wyświetlić wszystkie pliki w dół drzewa katalogowego od miejsca wywołania.

In [50]:
find

.
./Info 3 - Wykład 1.pdf
./Info 3 - Wykład 1.ipynb
./Zaznaczenie_004.png
./Info 3 - Wykład 2.ipynb
./burza_short
./burza
./fig2.png
./fig1.svg
./.ipynb_checkpoints
./.ipynb_checkpoints/Info 3 - Wykład 1-checkpoint.ipynb
./.ipynb_checkpoints/Info 3 - Wykład 2-checkpoint.ipynb
./Info 3 - Wykład 1.tex
./fig1.png
./unix.png
./Zaznaczenie_005.png
./quote-c-is-a-horrible-language-it-s-made-more-horrible-by-the-fact-that-a-lot-of-substandard-linus-torvalds-111-83-18.jpg


```find``` można wzbogacić o wiele opcji (patrz ```man```), najprzydatniejsze to:
* -maxdepth, -mindepth - określa zasięg (głębokość) działania
* -type określa typ pliku:
    * -type d - katalog,
    * -type l - dowiązanie,
    * -type f - zwykły plik
    * oraz kilka innych

np:

In [125]:
find ../../ -maxdepth 1 -mindepth 1 -type d

../../Info_CS_1
../../References
../../C++
../../Info_3
../../KAP_ACFD
../../Metody_Numeryczne
../../Mechanika_Plynow_2
../../Dyplomy
../../Mechanika_Plynow_1
../../Info_CS_2
../../MOMP_CFD


wyświetli tylko katalogi zaczynając od położenia w drzewie katalogowym dwa katalogi w górę (../../) z ograniczeniem głębokości przeszukiwania.

### Wyszukiwanie w tekście (ale nie tylko) - ```grep```
```grep``` - global regular expression print, drukuje linie, które zawierają wzorzec będący argumentem wywołania. Działanie ```grep``` pokażemy na przykładzie *Burzy* Williama Szekspira:

Pierwsze 10 linii pliku to:

In [59]:
head -n 10 burza

Burza
OSOBY:
ALONSO, król neapolitański.
SEBASTIAN, brat jego.
PROSPERO, prawy książę Mediolanu.
ANTONIO, brat jego, przywłaszczyciel księstwa.
FERDYNAND, syn króla neapolitańskiego.
GONZALO, poczciwy stary radca króla Neapolu.
ADRIAN, FRANCISKO, panowie.
KALIBAN, dziki i potworny niewolnik.


Wszystkie linie w których pojawia się *Ariel*:

In [63]:
grep Ariel burza

Przyjdź, Arielu[10]!
Wykona Ariel.
(Ariel wychodzi)
(wchodzi Ariel w postaci nimfy wodnej)
Piękne zjawisko! Lotny Arielu,
Wychodzi Kaliban — Wchodzi Ariel niewidzialny, gra i śpiewa, za nim Ferdynand.
Wzrok się ich zmierzył. Piękny Arielu,
(na stronie) Dobrześ się sprawił, Arielu! (głośno) Idźmy!
(do Ariela) A teraz słuchaj, co jeszcze masz zrobić!
Wchodzi Ariel niewidzialny, grając uroczystą melodię.
Zasypia — Ariel wychodzi.
Rozmawiają na stronie. Muzyka. Wchodzi Ariel niewidzialny.
Wchodzi Ariel niewidzialny.
Ariel gra na piszczałce i na bębenku.
Grzmoty i błyskawice. Wbiega Ariel w postaci harpii, roztacza skrzydła nad stołem; bankiet znika.
Udałeś pysznie harpię, Arielu,
Hej, Arielu, sługo mój przebiegły!
Lecz się nie zbliżaj, piękny Arielu,
Więc wszystko dobrze. — Przybądź Arielu!
Dzięki wam. — Arielu,
Jak myśli szybki, przybądź, Arielu!
Wchodzi Ariel.
(wchodzi Ariel, niosąc błyszczące materie itd.)
Prospero i Ariel zostają niewidzialni. — Wchodzą: Kaliban, Stefano i Trynkulo, pr

```grep``` jest szczególnie przydatny w połączeniu przez przekierowanie ```|``` z poleceniami ```ls``` lub ```find```.  
Możemy np. wylistować jedynie pliki .png:

In [67]:
ls -l |grep png

-rw-r-x--- 1 sgepner sgepner   13182 mar  3 16:24 fig1.png
-rw-rw-r-- 1 sgepner sgepner   16358 mar  3 16:24 fig2.png
-rw-r--r-- 1 sgepner sgepner 1166196 mar  3 17:37 unix.png
-rw-r--r-- 1 sgepner sgepner    6150 mar  3 18:03 Zaznaczenie_004.png
-rw-r--r-- 1 sgepner sgepner    1768 mar  3 18:03 Zaznaczenie_005.png


Albo przeszukać drzewo katalogowe na żądaną głębokość:

In [68]:
find | grep png

./Zaznaczenie_004.png
./fig2.png
./fig1.png
./unix.png
./Zaznaczenie_005.png


Możemy np. wykonać operację na wszystkich plikach pasujących to wzorca (też rozsianych po katalogach). powiedzmy, że chcemy usunąć wszystkie pliki png:

```rm -maxdepth 3 `find | grep png` ```

**Przykład z życia:** prowadzę obecnie obliczenia i generuję sporo plików z wynikami, które akurat nie są mi potrzebne. Ze względu na ograniczenie przestrzeni dyskowej muszę pilnować aby nie przekroczyć kwoty dyskowej (i nie narazić się na gniew admina). mam więc ustawioną pętlę (pętle w Bash omówimy za chwilę) w której wywołuję:  
```rm -maxdepth 3 `find | grep chk; sleep 600```  
aby co 10 minut posprzątać drzewo katalogowe z nieprzydatnych wyników.

#### Wyrażenia regularne
Skorzystamy z okazji zabaw z ```grep``` i wspomnimy o wyrażeniach regularnych. Są to operacje, podobne do arytmetycznych pozwalające na znalezienie ciągów znaków, odpowiadających zapisanemu wyrażeniu. Wyrażenia definiuje się według określonych zasad, i tak:

* Większość znaków odpowiada samym sobie (a,b, ..).
* Operator nawiasów kwadratowych '['  ']' pozwala na dopasowanie do dowolnego znaku, a użyty z '^' dopasowuje do znaków z poza listy.
* zakres znaków ustala się przez '-', np. 1-9 oznacza cyfry od 1 do 9.
* '^' i '$' samodzielnie oznaczają początek i koniec linii
* '.' - kropka, to dowolny znak

Przeszukajmy plik burza, ograniczając się do linii od 400 do 500 szukając fraz przy uzyciu wyrażeń regularnych. Szukamy fraz rozpoczynających się od spacji (literalizacja przez ```\```), litery *W* lub *w*, niezawierającej na trzecim miejscu *a* lub *e*, w końcu na ostatnim miejscu ma się znaleźć litera z zakresu *a-h*.

In [107]:
head -n 500 burza | tail -n 100 | grep \ [Ww][^ae][a-h]

I własnym kłamstwom wierzy — on uwierzył,
Zamierzył władzę i tytuł przywłaszczyć.
Dość wielkim państwem książki były dla mnie;
Spragniony władzy, z królem Neapolu
Przyrzeka z księstwa mnie i moich wygnać,


## Archiwizacja danych
Istnieje kilka narzędzi umożliwiających 'pakowanie' lików. Najpopularniejszym zdaje się być ```tar``` nazwa pochodzi of *tape archiver*, czyli archiwizacji na napędach taśmowych. Program wywołuje się wraz z opcjami. My ograniczymy się do:

* c - stwórz archiwum
* v - wyświetla pliki do pakowania / rozpakowania
* f - nazwa archiwum podana jako argument
* z - dodatkowo kompresuje archiwum
* x - rozpakuj archiwum

In [98]:
tar cvfz aaa.tar.gz fig1.png

fig1.png


In [101]:
tar xvf aaa.tar.gz

fig1.png


In [99]:
ls -l 

total 3256
-rw-r--r-- 1 sgepner sgepner   20480 mar  5 13:28  aaa.tar
-rw-r--r-- 1 sgepner sgepner   20480 mar  5 13:28  aaa.tar.gz
-rw-rw-r-- 1 sgepner sgepner  110559 mar  5 12:28  burza
-rw-rw-r-- 1 sgepner sgepner   13182 mar  3 16:24  fig1.png
-rw-rw-r-- 1 sgepner sgepner   17000 mar  3 16:23  fig1.svg
-rw-rw-r-- 1 sgepner sgepner   16358 mar  3 16:24  fig2.png
-rw-r--r-- 1 sgepner sgepner    2920 mar  3 18:04 'Info 3 - Wykład 1.aux'
-rw-r--r-- 1 sgepner sgepner  231815 mar  5 13:26 'Info 3 - Wykład 1.ipynb'
-rw-r--r-- 1 sgepner sgepner   35451 mar  3 18:04 'Info 3 - Wykład 1.log'
-rw-r--r-- 1 sgepner sgepner     699 mar  3 18:04 'Info 3 - Wykład 1.out'
-rw-r--r-- 1 sgepner sgepner 1227271 mar  3 18:04 'Info 3 - Wykład 1.pdf'
-rw-r--r-- 1 sgepner sgepner   21745 mar  3 18:04 'Info 3 - Wykład 1.synctex.gz'
-rw-rw-r-- 1 sgepner sgepner   24086 mar  3 18:04 'Info 3 - Wykład 1.tex'
-rw-rw-r-- 1 sgepner sgepner   24814 mar  3 17:58 'Info 3 - Wykład 1.tex~'
-rw-rw-r-- 1 sgepner sgepner 

## Procesy i zadania
Powłoka ```bash``` oferuje duże możliwości kontroli nad zadaniami użytkownika. Nas interesować będzie wyświetlanie działających procesów: ```ps``` (process show). Kończenie procesów przez ```kill```.Przenoszenie zadań w stan uśpienia przez kombinacje Ctrl-Z, kontrola uśpionych zadań ```jobs``` oraz wizualizacja drzewa procesów ```pstree```.

Zaczniemy od ```ps``` (flaga ```-f``` dodaje kilka szczegółów), który w podstawowej wersji wyświetla procesy działające w bieżącej sesji konsoli:

In [136]:
ps -f

UID        PID  PPID  C STIME TTY          TIME CMD
sgepner   8012 16766  0 14:38 pts/5    00:00:00 ps -f
sgepner  16766 16755  0 mar17 pts/5    00:00:00 /bin/bash --rcfile /home/sgepner


Omówmy najważniejsze elementy wyniku:

```UID        PID  PPID  C STIME TTY          TIME CMD
sgepner  10422 16766  0 16:22 pts/5    00:00:00 ps -f```

* UID - nazwa użytkownika będącego właścicielem procesu
* PID - Numer identyfikacyjny procesu
* PPID- Numer identyfikacyjny procesu nadrzędnego - rodzica
* C - wykorzystanie procesora prze proces
* STIME - Czas uruchomienia procesu
* TTY - Terminal procesu
* TIME - Całkowity czas wykorzystania procesora
* CMD - Polecenie, które wywołało proces

Inna przydatna flaga to ```-e``` - pozwala na wyświetlenie wszystkich procesów działających w systemie. Ponieważ może być ich bardzo dużo warto użyć razem z ```|grep```. Wyświetlmy wszystkie procesy użytkownika ```sgepner``` zawierające w sobie frazę ```bash```. Dodatkowo wynik ograniczymy do 4 linii:

In [142]:
ps -ef |grep sgepner |grep bash | head -n 4

sgepner    829  5599  0 mar14 pts/4    00:00:00 /bin/bash
sgepner   5600  5599  0 lut11 pts/1    00:00:00 /bin/bash
sgepner  11919 11918  0 14:50 pts/2    00:00:00 -bash
sgepner  16117 16766  0 15:04 pts/5    00:00:00 grep --color=auto bash


#### Znajdź i zabij (kill)
Zdarza się, że jeden z procesów przestanie odpowiadać - zawiesi się. Albo po prostu nie chcemy go widzieć w puli procesów. Z pomocą przychodzi nam polecenie ```kill``` najczęściej wywoływane z opcją ```-9```. W celu określenia jaki proces chcemy "zkilować" musimy znać jego PID (patrz ```ps```). Powiedzmy, że uśmiercić chcemy proces powłoki o numerze PID=829, piszemy wówczas:

``` kill 829 ```

lub dla większej pewności powodzenia

```kill -9 829 ```

Różnica jest taka, że w pierwszym przypadku ```kill``` wysyła do procesu sygnał SIGTERM, czyli uprzejmą prośbę o zakończenie działania i zwolnienie zasobów. Flaga ```-9``` odpowiada SIGKILL oznacza natychmiastowe zakończenie działania.

Ciekawym rozwinięciem jest ```killall``` pozwala na zabicie wszystkich procesów wywołanych konkretnym poleceniem. Np. jeżeli chcemy wyłączyć wszystkie uruchomione sesje ```autossh``` (utrzymuje persystentne połączenie, np. w celu podtrzymywania tunelu *ssh*) możemy zrobić tak:

In [145]:
ps -e |grep autossh

 2101 ?        00:00:01 autossh
 3851 ?        00:00:00 autossh


```kill -9 2101 3851```

albo

```killall autossh```

#### Drzewo procesów
Procesy powoływane sa do życia w sposób hierarchiczny. Tj. każdy proces wywodzi się z procesu rodzica, stąd numer PPID. Pierwszym procesem, jest ```init``` lub w systemie, na którym przygotowywany jest ten dokument ```systemd``` ([Dla zainteresowanych trochę więcej szczegółów pod tym linkiem.](https://en.wikipedia.org/wiki/Linux_startup_process#SysV_init))

Drzewo procesów można wyświetlić używając ```pstree```. Jak zwykle dostępne jest wiele opcji (```man pstree```). My ograniczymy się do: 
* -p PID - pokaz proces o numerze PID
* -T nie pokazuj wątków
* -s pokaż rodziców

In [155]:
pstree -p 5599 -T -s

systemd(1)───systemd(1589)───screen(5599)─┬─bash(829)
                                          ├─bash(5600)───jupyter-noteboo(15339)+++
                                          └─bash(18197)───top(21888)


#### Kontrola procesów użytkownika
Użytkownik, mając do dyspozycji powłokę może kontrolować wiele procesów. Zadania które wywołuje mogą działać 'w tle' (background) albo 'na wierzchu' (foreground). Na wierzchu może być tylko jedno zadanie, podczas gdy w tle działać może ich wiele. Dodatkowo działające zadania można przenosić w stan uśpienia przy użyciu kombinacji *Ctrl-Z*, przywoływać na pierwszy plan komendą ```fg```, odsyłać do pracy w tle używając ```bg``` czy w końcu zabić kombinacją *Ctrl-C* bądź ```kill```.

By zadanie przenieść do tła zaraz po wywołaniu na jego końcu dodaje się ```&```. Do przykładu potrzebujemy procesu, który będzie działał odpowiednio długo, przeszukamy więc katalogi użytkowników i katalog systemowy, a wynik wraz ze wszystkimi błędami (```>&```) przekierujemy do plików:

In [165]:
find / 1> result_file 2> /dev/null&
find /home >& result_file2&

[1] 1808
[2] 1809


Wywołaliśmy dwa zadania (jobs), ```ps``` pokaże co następuje:

In [166]:
ps 

  PID TTY          TIME CMD
 1808 pts/5    00:00:00 find
 1809 pts/5    00:00:00 find
 1879 pts/5    00:00:00 ps
16766 pts/5    00:00:00 bash


Możemy wyświetlić bieżące zadania poleceniem ```jobs```

In [167]:
jobs

[1]-  Running                 find / > result_file 2> /dev/null &
[2]+  Running                 find /home &> result_file2 &


W środowisku Jupiter nie mamy dostępu do kombinacji klawiszy *Ctrl-Z* i *Ctrl-C* zawieszanie i zabijanie procesów należy przećwiczyć we własnym zakresie.

Możemy natomiast użyć ```kill``` wraz z identyfikatorem zadania (nie numerem PID):

In [168]:
kill %2

: 1

In [169]:
jobs

[1]+  Running                 find / > result_file 2> /dev/null &


Albo podając PID:

In [170]:
kill 1808

: 1

In [172]:
jobs

In [173]:
ps

  PID TTY          TIME CMD
 4004 pts/5    00:00:00 ps
16766 pts/5    00:00:00 bash


Oba zadania zostały zakończone.

## Skrypty BASH

TBA